File: python/opendesk/protocol/transports/websocket.py
"""WebSocket transport for the opendesk wire protocol.

Wraps the ``websockets`` library so each opendesk protocol frame is one
WebSocket binary message.  Text frames are rejected — we are a strictly
binary protocol.

Client usage::

    from opendesk.protocol.transports.websocket import connect_websocket
    from opendesk.computer.remote import RemoteComputer

    conn = await connect_websocket("ws://mac-mini.local:8423")
    remote = await RemoteComputer.connect(conn)

Server usage::

    from opendesk.protocol.transports.websocket import serve_websocket

    async def handle(conn):
        # build a Peer / dispatcher around `conn` and run the session
        ...

    async with serve_websocket(handle, host="0.0.0.0", port=8423) as server:
        await server.wait_closed()
"""

from __future__ import annotations

import asyncio
import contextlib
from typing import Awaitable, Callable, Optional

try:
    import websockets
    from websockets.asyncio.client import ClientConnection
    from websockets.asyncio.server import Server, ServerConnection
    from websockets.exceptions import ConnectionClosed as _WSClosed
except ImportError as _exc:  # pragma: no cover
    raise ImportError(
        "websockets is required for the WebSocket transport. "
        "Install with: pip install 'opendesk[remote]'"
    ) from _exc

from opendesk.protocol.connection import Connection, ConnectionClosed


# A connection can be either a server-side ``ServerConnection`` or a
# client-side ``ClientConnection`` — both expose the same send/recv/close API.
_WSConnection = "ServerConnection | ClientConnection"


class WebSocketConnection(Connection):
    """One opendesk :class:`Connection` backed by a single WebSocket link."""

    def __init__(self, ws) -> None:
        self._ws = ws
        self._closed = False

    async def send(self, data: bytes) -> None:
        if self._closed:
            raise ConnectionClosed("send on closed connection")
        try:
            await self._ws.send(data)
        except _WSClosed as exc:
            self._closed = True
            raise ConnectionClosed(str(exc)) from exc

    async def recv(self) -> bytes:
        if self._closed:
            raise ConnectionClosed("recv on closed connection")
        try:
            msg = await self._ws.recv()
        except _WSClosed as exc:
            self._closed = True
            raise ConnectionClosed(str(exc)) from exc
        if isinstance(msg, str):
            await self.aclose()
            raise ConnectionClosed(
                "received text frame; opendesk protocol is binary-only"
            )
        return msg

    async def aclose(self) -> None:
        if self._closed:
            return
        self._closed = True
        with contextlib.suppress(Exception):
            await self._ws.close()


# ---------------------------------------------------------------------------
# Client
# ---------------------------------------------------------------------------


async def connect_websocket(
    url: str,
    *,
    ssl=None,
    additional_headers: Optional[dict] = None,
    max_message_size: Optional[int] = None,
) -> WebSocketConnection:
    """Open a WebSocket :class:`Connection` to *url*.

    ``url`` is ``ws://`` or ``wss://``.  ``max_message_size=None`` (default)
    removes the per-frame size cap — needed for full-resolution display
    captures.
    """
    ws = await websockets.connect(
        url,
        ssl=ssl,
        additional_headers=additional_headers,
        max_size=max_message_size,
        # We use msgpack ourselves; turn off WebSocket-level compression.
        compression=None,
    )
    return WebSocketConnection(ws)


# ---------------------------------------------------------------------------
# Server
# ---------------------------------------------------------------------------


ConnectionHandler = Callable[[WebSocketConnection], Awaitable[None]]


class WebSocketServer:
    """A running WebSocket server.  Use via :func:`serve_websocket`."""

    def __init__(self, server: Server, host: str, port: int) -> None:
        self._server = server
        self.host = host
        self.port = port

    async def wait_closed(self) -> None:
        await self._server.wait_closed()

    def close(self) -> None:
        self._server.close()

    async def aclose(self) -> None:
        self._server.close()
        await self._server.wait_closed()

    async def __aenter__(self) -> "WebSocketServer":
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        await self.aclose()


async def serve_websocket(
    handler: ConnectionHandler,
    *,
    host: str = "0.0.0.0",
    port: int = 8423,
    ssl=None,
    max_message_size: Optional[int] = None,
) -> WebSocketServer:
    """Start a WebSocket server and invoke *handler* for each connection.

    *handler* receives a :class:`WebSocketConnection`.  When it returns (or
    raises), the connection is closed.

    Pass ``port=0`` to bind to an ephemeral port — read the actual port back
    from the returned :attr:`WebSocketServer.port`.
    """
    async def _handler(ws) -> None:
        conn = WebSocketConnection(ws)
        try:
            await handler(conn)
        except _WSClosed:
            pass
        finally:
            await conn.aclose()

    server = await websockets.serve(
        _handler,
        host=host,
        port=port,
        ssl=ssl,
        max_size=max_message_size,
        compression=None,
    )

    # Resolve the actual port when port=0 was requested.
    actual_port = port
    socks = getattr(server, "sockets", None) or []
    for s in socks:
        try:
            actual_port = s.getsockname()[1]
            break
        except Exception:  # pragma: no cover
            continue

    return WebSocketServer(server, host=host, port=actual_port)

File: js/src/protocol/transports/websocket.ts
/**
 * WebSocket transport — mirrors Python opendesk.protocol.transports.websocket.
 *
 * Each opendesk protocol frame is one WebSocket binary message.
 * Uses the `ws` package.
 */

import WebSocket, { WebSocketServer as WsServer } from "ws";
import { Connection, ConnectionClosed } from "../connection.js";

interface Waiter {
  resolve: (data: Buffer) => void;
  reject: (err: Error) => void;
}

export class WebSocketConnection extends Connection {
  private ws: WebSocket;
  private queue: Buffer[] = [];
  private waiters: Waiter[] = [];
  private _closed = false;
  private _closeError?: Error;

  constructor(ws: WebSocket) {
    super();
    this.ws = ws;

    ws.on("message", (data) => {
      const buf = Buffer.isBuffer(data) ? data : Buffer.from(data as ArrayBuffer);
      if (this.waiters.length > 0) {
        this.waiters.shift()!.resolve(buf);
      } else {
        this.queue.push(buf);
      }
    });

    const onClose = () => {
      this._closed = true;
      const err = this._closeError ?? new ConnectionClosed("WebSocket closed");
      for (const w of this.waiters) w.reject(err);
      this.waiters = [];
    };

    ws.on("close", onClose);
    ws.on("error", (err) => {
      this._closeError = new ConnectionClosed(err.message);
      // close event follows error event; waiters drained there
    });
  }

  async send(data: Buffer): Promise<void> {
    if (this._closed) throw new ConnectionClosed();
    return new Promise((resolve, reject) => {
      this.ws.send(data, (err) => {
        if (err) reject(new ConnectionClosed(err.message));
        else resolve();
      });
    });
  }

  async recv(): Promise<Buffer> {
    if (this._closed && this.queue.length === 0) {
      throw this._closeError ?? new ConnectionClosed();
    }
    if (this.queue.length > 0) return this.queue.shift()!;
    return new Promise((resolve, reject) => {
      this.waiters.push({ resolve, reject });
    });
  }

  async close(): Promise<void> {
    if (this._closed) return;
    this._closed = true;
    this.ws.terminate();
    const err = new ConnectionClosed("local close");
    for (const w of this.waiters) w.reject(err);
    this.waiters = [];
  }
}

export async function connectWebSocket(url: string): Promise<WebSocketConnection> {
  return new Promise((resolve, reject) => {
    const ws = new WebSocket(url, { perMessageDeflate: false });
    ws.once("open", () => resolve(new WebSocketConnection(ws)));
    ws.once("error", (err) => reject(new Error(`WebSocket connect failed: ${err.message}`)));
  });
}

export type ConnectionHandler = (conn: WebSocketConnection) => Promise<void>;

export interface ServeOptions {
  host?: string;
  port?: number;
}

export class WebSocketServerWrapper {
  private server: WsServer;
  readonly port: number;
  readonly host: string;

  constructor(server: WsServer, host: string, port: number) {
    this.server = server;
    this.host = host;
    this.port = port;
  }

  async close(): Promise<void> {
    await new Promise<void>((resolve, reject) => {
      this.server.close((err) => (err ? reject(err) : resolve()));
    });
  }
}

export async function serveWebSocket(
  handler: ConnectionHandler,
  opts: ServeOptions = {},
): Promise<WebSocketServerWrapper> {
  const host = opts.host ?? "0.0.0.0";
  const port = opts.port ?? 8423;

  return new Promise((resolve, reject) => {
    const server = new WsServer({ host, port: port === 0 ? undefined : port });

    server.on("connection", async (ws) => {
      const conn = new WebSocketConnection(ws);
      try {
        await handler(conn);
      } catch {
        // ignore per-connection errors at server level
      } finally {
        await conn.close().catch(() => {});
      }
    });

    server.once("error", reject);
    server.once("listening", () => {
      const addr = server.address() as { port: number };
      resolve(new WebSocketServerWrapper(server, host, addr?.port ?? port));
    });
  });
}
